src/hotspot/share/oops/jmethodIDTable.cpp

```diff
@@ -0,0 +1,185 @@
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "logging/log.hpp"
+#include "memory/allocation.hpp"
+#include "memory/resourceArea.hpp"
+#include "oops/method.hpp"
+#include "oops/jmethodIDTable.hpp"
+#include "runtime/mutexLocker.hpp"
+#include "utilities/concurrentHashTable.inline.hpp"
+#include "utilities/concurrentHashTableTasks.inline.hpp"
+#include "utilities/macros.hpp"
+
+// Save (jmethod, Method*) in a hashtable to lookup Method
+// The CHT is for performance because it is has lock free lookup.
+
+static uint64_t _jmethodID_counter = 0;
+
+class JmethodEntry {
+ public:
+  uint64_t _id;
+  Method*  _method;
+
+  JmethodEntry(uint64_t id, Method* method) : _id(id), _method(method) {}
+};
+
+class JmethodIDTableConfig : public AllStatic {
+ public:
+  typedef JmethodEntry Value;
+  static void* allocate_node(void* context, size_t size, Value const& value) {
+    return AllocateHeap(size, mtJNI);
+  }
+  static void free_node(void* context, void* memory, Value const& value) {
+    FreeHeap(memory);
+  }
+  static uintx get_hash(Value const& value, bool* is_dead) {
+    *is_dead = false;
+    return value._id;
+  }
+  static bool is_dead(Value const& value) {
+    return false;
+  }
+};
+
+using MethodIdTable = ConcurrentHashTable<JmethodIDTableConfig, mtJNI>;
+static MethodIdTable* _jmethod_id_table = nullptr;
+
+void JmethodIDTable::initialize() {
+  const size_t start_size = 10;
+  // 2^24 is max size
+  const size_t end_size = 24;
+  // If a chain gets to 32 something might be wrong
+  const size_t grow_hint = 32;
+
+  _jmethod_id_table  = new MethodIdTable(start_size, end_size, grow_hint);
+}
+
+class JmethodIDLookup : StackObj {
+ private:
+  uint64_t _mid;
+
+ public:
+  JmethodIDLookup(const uint64_t mid) : _mid(mid) {}
+  uintx get_hash() const {
+    return _mid;
+  }
+  bool equals(JmethodEntry* value) {
+    return _mid == value->_id;
+  }
+
+  bool is_dead(JmethodEntry* value) {
+    return false;
+  }
+};
+
+static JmethodEntry* get_jmethod_entry(jmethodID mid) {
+  assert(mid != nullptr, "JNI method id should not be null");
```

Comment:
Perhaps: s/null/nullptr/
I can't remember if assert failure text output is okay to be `null`.